app/models/action.rb

```diff
@@ -0,0 +1,38 @@
+class Action < ActiveRecord::Base
+  TERMINAL_STATUSES = %w(done failed)
+
+  belongs_to :resource, polymorphic: true
+
+  attr_accessible :initial_eta, :resource
+
+  def self.enclose!(initial_eta, klass, method, *args)
+    action = create!(initial_eta: initial_eta)
+    ActionWorker.perform_async(action.id, klass.to_s, method, *args)
+    action
+  end
+
+  def eta(time = Time.now)
+    diff = (created_at + initial_eta) - time
```

Comment:
I'm just beginning to read the changes, so I don't know about the mechanics, but here's what popped into my mind:

If sidekiq is involved and the actiom model is created synchronously, the job could not start until after the eta is reached. Maybe we should use `updated_at` and `touch` the action as soon as we start processing the job.